Fix : Snackbar Notice Inconsistency

Vrishabhsk opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Fixes: https://github.com/WordPress/gutenberg/issues/66364

<!-- In a few words, what is the PR actually doing? -->
- Enable `View Options` other than `Distraction Fee` to show a `snackbar notice` when they are `toggled`
- `View Options` affected : `Top toolbar`  `Spotlight mode`  `Fullscreen mode`

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
- https://github.com/WordPress/gutenberg/issues/66364
> Either all these options should trigger a snackbar notice or none.
I'd like to propose to remove the snackbar notice of the Distraction free mode for consistency and simplicity. I'm not sure this snackbar notice adds much value.

> Note: recommended to check what happens when toggling these options via the Command palette.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- Create toggle actions similar to `toggleDistractionFree` for other `View Options`
- Replace hardcoded functionality with these `toggles` which allows `Snackbar Notice` to be visible when these options are toggled via the `More Menu`

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
1. Edit a post.
2. Go to the top bar > Options.
3. Toggle on and off each view option.
4. Observe all modes triggers a snackbar notice.

Vrishabhsk committed:
Add actions for toggling view options with a Notice

Vrishabhsk committed:
Modify Menu Item components and commands to use new toggles

ntsekouras commented on packages/edit-post/src/components/more-menu/index.js:
```diff
@@ -14,12 +14,17 @@ import { unlock } from '../../lock-unlock';
 const MoreMenu = () => {
 	const isLargeViewport = useViewportMatch( 'large' );
 
+	// Action to toggle the FullscreenMode with a Notice for visual feedback.
```
Nit: no need for this comment I think.

ntsekouras commented on packages/edit-post/src/components/more-menu/index.js:
```diff
@@ -14,12 +14,17 @@ import { unlock } from '../../lock-unlock';
 import ManagePatternsMenuItem from './manage-patterns-menu-item';
 import WelcomeGuideMenuItem from './welcome-guide-menu-item';
 import EditPostPreferencesModal from '../preferences-modal';
+import { store as editorStore } from '../../store';
```
I was so confused at this first... This is the `editPostStore`, so you need to rename.

ntsekouras commented on packages/edit-post/src/components/more-menu/index.js:
```diff
@@ -29,6 +34,8 @@ const MoreMenu = () => {
 						name="fullscreenMode"
 						label={ __( 'Fullscreen mode' ) }
 						info={ __( 'Show and hide the admin user interface' ) }
+						handleToggling={ false }
```
This prop existed from before, but it's so weird..

Vrishabhsk replied:
This will be required since `handleToggling` prop has default set to `true` and will cause `ambiguity` during `toggling`

ntsekouras replied:
It just shouldn't be there in a component that has `toggle` in its name `PreferenceToggleMenuItem` :)

ntsekouras commented on packages/edit-post/src/store/actions.js:
```diff
@@ -509,3 +511,45 @@ export const toggleDistractionFree =
+
+/**
+ * Action that toggles the Fullscreen Mode view option.
+ * Shows a notice when option is toggled.
```
Nit: I think we can skip the `Shows a notice when option is toggled` in all new actions.

ntsekouras commented on packages/edit-post/src/store/actions.js:
```diff
@@ -509,3 +511,45 @@ export const toggleDistractionFree =
+export const toggleFullscreenMode =
+	() =>
+	( { registry } ) => {
+		// Determine the current state of the Fullscreen Mode option.
```
Nit: no need for these comments in the whole function.

ntsekouras commented on packages/editor/src/store/actions.js:
```diff
@@ -873,6 +873,82 @@ export const toggleDistractionFree =
+export const toggleSpotlightMode =
+	() =>
+	( { registry } ) => {
+		// Determine the current state of the Spotlight Mode option.
```
Nit: same as above regarding the removal of these comments..

ntsekouras commented on packages/editor/src/components/commands/index.js:
```diff
@@ -162,28 +145,8 @@ function useEditorCommandLoader() {
 		label: __( 'Toggle top toolbar' ),
 		callback: ( { close } ) => {
-			toggle( 'core', 'fixedToolbar' );
-			if ( isDistractionFree ) {
```
You didn't include this in the new action, but it seems you're correct as I couldn't find any reason for doing that.. It seems as a leftover, but I'll cc @draganescu to make sure.

Vrishabhsk replied:
- This functionality was already implemented in the `actions.js` so I removed it from here
- There is a dependency between `Top toolbar` and `Distraction Free` view options
-  I have handled and tested all `possible combinations` for the `trunk` and `fix/options-snackbar` branch

ntsekouras reviewed (commented):
Thanks for the PR, it looks good! My comments are mostly nits about inline comments :).

The thing that is not handled here though is the duplicate speak message. I think for these use cases it makes most sense to just use `MenuItem` component directly and avoid the weird `handleToggling` and `messageActivated/messageDeactivated` messages.

Vrishabhsk committed:
Allow custom speak message

Vrishabhsk committed:
Remove comments and add custom speak message

Vrishabhsk committed:
Allow disabling speak fn call in PreferenceToggleMenuItem component

Vrishabhsk committed:
Remove speak messages. Add disableSpeech prop

Vrishabhsk commented:
Hi @ntsekouras 👋. Thanks for the detailed review
- I have added a couple of changes to fix the duplicate `speak` message issue
- Added `disableSpeech` prop to `PreferenceToggleMenuItem component` to prevent `speakMessage()` from executing
- Added `speakMessage` option to the `createNotice` function called by `createInfoNotice` for custom message capability

Let me know your thoughts on the changes. Thanks

Vrishabhsk commented:
To confirm and validate the `A11y Speak` message, you can use the following script : [`a11y-speak-monitor.js`](https://codefile.io/f/OQBymdlwjD)

ntsekouras commented on packages/preferences/src/components/preference-toggle-menu-item/index.js:
```diff
@@ -23,6 +23,7 @@ export default function PreferenceToggleMenuItem( {
 	handleToggling = true,
 	onToggle = () => null,
 	disabled = false,
+	disableSpeech = false,
```
We should be conservative when adding props to components and especially public ones, like this one. My comment related to `handleToggling` was about the same thing - it shouldn't be added in the first place.

>it makes most sense to just use MenuItem component directly and avoid the weird handleToggling and messageActivated/messageDeactivated messages.

`PreferenceToggleMenuItem` in essence is a thin wrapper of `MenuItem` for toggling preferences and speak the messages. In these cases we handle both of these in the action, so there is no need to use `PreferenceToggleMenuItem` at all, and instead use `MenuItem` directly.

Vrishabhsk replied:
Hi @ntsekouras 👋 Thanks for review
- I have addressed the changes in the latest commit

Let me know your thoughts on it. Thanks

Vrishabhsk committed:
Address Feedback

Replace PreferenceToggleMenuItem with MenuItem. Replace function called when Fullscreen Mode shortcut is used to show the notice. Fix speakMessage placement. Remove disableSpeech prop from PreferenceToggleMenuItem component

ntsekouras commented on packages/notices/src/store/actions.js:
```diff
@@ -69,6 +71,7 @@ let uniqueId = 0;
 export function createNotice( status = DEFAULT_STATUS, content, options = {} ) {
 	const {
 		speak = true,
+		speakMessage = null,
```
I don't get why we would need this - at least in this case. We shouldn't have two different messages.

Vrishabhsk replied:
- The messages used for the `PreferenceToggleMenuItem` were different from the `label` used, which is why I implemented this
- If we remove this for the `Menu Items`, the message would be `Distraction Free off / Distraction Free on` which does not provide much context.
- Although for argument's sake, even `Distraction Free mode activated` also does not provide much context. 
- I would suggest expanding this custom message to give more context for e.g. : `Distraction Free view option is enabled/disabled for your editor`

ntsekouras replied:
>Although for argument's sake, even Distraction Free mode activated also does not provide much context.

Yeah, I don't think we would need two different messages for now.

ntsekouras replied:
Let's please remove this `speakMessage` prop. We don't need it as it already speaks the content.

ntsekouras reviewed (commented):
Thank you for iterating!

Related though to the [latest feedback ](https://github.com/WordPress/gutenberg/issues/66364#issuecomment-2439963931)on the issue, maybe it's better **not** to show a notice on the more menu clicks, but also preserve the spoken messages.

So some refactoring will be needed. I think by adding a `createNotice` param (with default `true) in `toggleDistractionFree` and revert to use `PreferenceToggleMenuItem` in the more menu items would be enough.

Vrishabhsk commented:
So just that we are on the same page :
- Add `createNotice` param in `toggleDistractionFree` function
- Revert `MenuItem` back to `PreferenceToggleMenuItem` and remove the need to use `onToggle` as it was setup initially
- Leave the newly created `actions` and `notice` related changes as they are for now

Vrishabhsk commented:
I also noticed quite a few changes made to the labels here : https://github.com/WordPress/gutenberg/pull/66371

ntsekouras commented:
>So just that we are on the same page :

>Add createNotice param in toggleDistractionFree function
Revert MenuItem back to PreferenceToggleMenuItem and remove the need to use onToggle as it was setup initially
Leave the newly created actions and notice related changes as they are for now

That were my first thoughts but didn't time in code. Probably it will be enough.

To elaborate a bit.. 
1. The new actions are needed to use in context of commands etc.. They just toggle a single preference and speak the message.
2. `PreferenceToggleMenuItem` handles a preference toggling as is and handles speaking a message. So in the more menu items we don't need to dispatch the action. We need though the messages to match the ones in the actions.
3. `toggleDistractionFree` is already an action and is also triggered through other user actions. Since we have extra logic in that one, we should probably have a param(createNotice) in the action not to create a notice in the more menu item.

Makes sense?

Vrishabhsk committed:
Revert PreferenceToggleMenuItem and add createNotice param for toggleDistractionFree action

Vrishabhsk committed:
Merge branch 'trunk' into fix/options-snackbar

ntsekouras commented on packages/editor/src/components/more-menu/index.js:
```diff
@@ -30,11 +30,13 @@ export default function MoreMenu() {
-			select( preferencesStore ).get( 'core', 'showIconLabels' ),
-		[]
-	);
+	const { showIconLabels } = useSelect( ( select ) => {
```
Is this a leftover from other commits? Nothing really changed here.

ntsekouras commented on packages/editor/src/store/actions.js:
```diff
@@ -837,42 +839,122 @@ export const toggleDistractionFree =
+													'core',
+													'fixedToolbar',
+													isDistractionFree
+														? true
```
That was from before, but let's just remove the ternary operator and keep `isDistractionFree`.

ntsekouras commented on packages/editor/src/store/actions.js:
```diff
@@ -812,9 +812,11 @@ export function setIsListViewOpened( isOpen ) {
  */
 export const toggleDistractionFree =
-	() =>
+	( createNotice = true ) =>
```
```suggestion
	( { createNotice = true } = {} ) =>
```

Let's make this an object to be safer (in case of future additions) and have more clarity about the params we pass.

ntsekouras reviewed (commented):
This is really close, thank you!

Besides the comments I left, we need to do:
1. Remove `scope and name` [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/more-menu/index.js#L76) as we don't use them. Also pass the new param at `onToggle` like this: `toggleDistractionFree( { createNotice: false } )`. 
2. We should have consistent messages in both more menus and actions. That means either have `${option} off/on` or `${option activated/deactivated`. I'd go for `on/off` personally.
3. Commented about it, but very important to remove the `speakMessage` prop you added here.

With these addressed, I think we can land.

Vrishabhsk committed:
Feedback changes

Ensure speak messages are consistent for actions and menu items. Remove speakMessage option for createInfoNotice. Revert extraction of showIconLabels. Convert createNotice param to an object based param.

Vrishabhsk committed:
Fix createNotice docs

Vrishabhsk committed:
Disable notice for Distraction Free Menu Item toggle

Vrishabhsk committed:
Fix incorrect activation notice bug

Vrishabhsk commented:
Thanks @ntsekouras for the review 👍 
- I went ahead with `${option} activated/deactivated.` to keep the messages consistent.
- Disabled the notice for `Distraction Free` Menu Item
- We cannot remove `scope and name` since it determines its `isActive` attribute which renders the `check` icon
- I have removed the `speakMessage` prop

ntsekouras reviewed (approved):
Great work here, thanks!
